### PR TITLE
Process block data programs early

### PR DESCRIPTION
**Pull request by vdonaldson**
Block data programs may initialize common block data, so process them
in a preliminary pass (along with function declarations and module
variables) to give them precedence over other common block declarations.

### CODE REVIEW

File: flang/test/Lower/array.f90 [hunk 1]
```diff
@@ -76,6 +76,26 @@ subroutine s(i,j,k,ii,jj,kk,a1,a2,a3,a4,a5,a6,a7)
   
 end subroutine s
 
+! CHECK: %[[VAL_1:.*]] = constant 1.000000e+00 : f32
```
Comment (jeanPerier) [inline flang/test/Lower/array.f90:79]: Could you add a CHECK-LABEL with the common block `fir.global  someName` here. Now that these checks are placed before the block data code, they are a bit weird to read out of any context.

**Review by jeanPerier — APPROVED** [review]
Looks OK to me. This won't cover the case where common are initialized outside of block data, but at least should cover conformant programs. We may want to either forbid common initialization outside of block data in semantics like NAG does or improve support here later.

**Review by psteinfeld — COMMENTED** [review]
I don't understand the code, but I can verify that everything builds, tests, and is formatted correctly.

File: flang/lib/Lower/Bridge.cpp [hunk 2]
```diff
@@ -106,34 +106,47 @@ using IncrementLoopNestInfo = llvm::SmallVector<IncrementLoopInfo>;
 
 namespace {
 
-/// Walk over the pre-FIR tree (PFT) and lower it to the FIR dialect of MLIR.
-///
-/// After building the PFT, the FirConverter processes that representation
-/// and lowers it to the FIR executable representation.
+/// Create a pre-FIR tree (PFT) from the parse tree, then traverse it to
```
Comment (schweitzpgi) [inline flang/lib/Lower/Bridge.cpp:109]: I don't think FirConverter actually creates the PFT though. Has that been changed?

It does traverse the PFT and lower to IR.

**Review by schweitzpgi — COMMENTED** [review]
thanks for the fix, Val.

File: flang/lib/Lower/Bridge.cpp (hunk 2 above)
Comment (vdonaldson) [reply flang/lib/Lower/Bridge.cpp:109]: I'll change this to _Traverse the pre-FIR tree (PFT) to generate the FIR dialect of MLIR._

**Pull request merged**